JOSS Paper (#1)
ottrproject/metricminer: paper.md

# Statement of need

Previous research by our group showed that the main reason researchers do not collect software metrics is a lack of time and funding to do so [@bioinformatics]. This creates a compounding missed opportunity: metrics can help developers learn more about how their software is used and can help demonstrate impact to funders which would, in turn, give developers more insights and funding respectively to increase the software's impact. Yet, the lack of time to collect metrics means developers can't fully learn about their software or quickly communicate this to funders. Collecting meaningful metrics from Application Programming Interfaces (APIs) can be time consuming, so `metricminer` aims to be an open-source solution to make the metric collection process more efficient.

> `metricminer` is an R package designed to simplify the process of collecting metrics from common web services. R provides an ideal environment for data analysis and visualization, while allowing for reproducible research workflows. `metricminer` was designed to provide a consistent and user-friendly interface to various web services including GitHub, Google Analytics, Calendly, Google Forms, and YouTube. The package handles authentication, data retrieval, and transformation of the raw API responses into opinionated tidy data frames, making it immediately useful for analysis [@wickham_tidy_2014].

Reviewer: Not sure about the word "opinionated" in the final sentence. Thinking you might want something like "standardized" or "propagated"

Reviewer: From Carrie: Could possibly even remove that word and keep the phrase without an adjective since it is just a tidy data frame

Author: I do mean opinionated. In that a meaningful dataset you have to make opinions/decisions about. Intending to use it in the way its used here: https://www.researchgate.net/publication/341462829_Opinionated_Data_Analysis

Author: But if you both don't like it, then I don't have to die on this hill, you can change it no problem.

Author: Opinion as in: there is no way for us to know exactly what data someone might want from an API return. But we used our own opinions to inform what data we would return. It's not a standard.

Reviewer: I meant standardized as in any project that I'm working on with metricminer will have predictable, standardized data frames in comparison to each other. I better understand what you meant by opinionated now, but I think a whole sentence would be necessary to clearly communicate that to a broad audience. Not sure how I'd phrase the sentence though.

Reviewer 2: maybe something like: 
The package handles authentication, data retrieval, and transformation of the raw API responses into tidy data frames, making it immediately useful for analysis [@wickham_tidy_2014]. The data selected are based on our opinions of what has been useful for evaluating the impact of our software, scientific papers, and training outreach programs.

Reviewer 2: Please ignore and see my comment below.

Reviewer 2: going to resolve this so we just focus on the suggested change below.

> `metricminer` is an R package designed to simplify the process of collecting metrics from common web services. R provides an ideal environment for data analysis and visualization, while allowing for reproducible research workflows. `metricminer` was designed to provide a consistent and user-friendly interface to various web services including GitHub, Google Analytics, Calendly, Google Forms, and YouTube. The package handles authentication, data retrieval, and transformation of the raw API responses into opinionated tidy data frames, making it immediately useful for analysis [@wickham_tidy_2014].

Author: ```suggestion
`metricminer` is an R package designed to simplify the collection of metrics from common web services. R provides an ideal environment for data analysis and visualization while supporting reproducible research workflows. `metricminer` offers a consistent, user-friendly interface to a variety of web services, including GitHub, Google Analytics, Calendly, Google Forms, and YouTube. The package handles authentication, data retrieval, and transformation of raw API responses into tidy data frames, making results immediately useful for downstream analysis [@wickham_tidy_2014].

Because metric collection inherently involves design decisions, the tidy data returned by `metricminer` is deliberately opinionated: it specifically grabs portions of the API responses that we found most valuable for evaluating our scientific software, papers, and training programs. A consequence of this approach is that all `metricminer` users receive a comparable, standardized dataset, facilitating shared understanding and cross-project comparisons.
```

Author: See how you feel about this @kweav

Reviewer: I like that a lot better! 
Two small notes:
* There's one "meticminer" without the backticks at the start of the sentence "metricminer offers a ..."
* "surfacing only what is most relevant to the user" --> "surfacing only what is likely to be relevant for users tracking resources of these types" ... or something like that since we know what has been relevant for us (and we're often users), but we can't possibly know what will be most relevant for every user. Trying to keep it general and not say "we" but also wanting to keep the heart of what you're saying because I like it a lot!

Reviewer 2: oh sorry I missed this but offered a suggestion above... maybe we can merge the too based on Kate's comments too?

Reviewer 2: so something like: 

Because metric collection inherently involves design decisions, the tidy data returned by `metricminer` is deliberately opinionated: it specifically grabs portions of the AI responses that we found most valuable  for evaluating our scientific software, papers, and training programs. A consequence of this approach is that all `metricminer` users receive a comparable, standardized dataset, facilitating shared understanding and cross-project comparisons.

Reviewer: I like your suggestion @carriewright11 and think it reflects what @cansavvy was communicating too!

Author: That works!

Reviewer 2: OK I updated your suggestion

Reviewer: I also just updated it to add those backticks and then to change AI --> API. Didn't realize we could edit or I would have handled the backticks earlier :)

Reviewer: Good to commit that change?

> `metricminer` was developed to serve both researchers, educators, and other practitioners who need to collect and analyze metrics from multiple platforms. It has been specifically designed to support educational initiatives and research projects that require tracking engagement across various platforms. The combination of simplified authentication, consistent data formats, and integration with the R ecosystem makes `metricminer` particularly valuable for organizations that need to generate regular reports or maintain dashboards of their metrics.

Reviewer: ```suggestion
`metricminer` was developed to serve researchers, educators, and other practitioners who need to collect and analyze metrics from multiple platforms. It has been specifically designed to support educational initiatives and research projects that require tracking engagement across various platforms. The combination of simplified authentication, consistent data formats, and integration with the R ecosystem makes `metricminer` particularly valuable for organizations that need to generate regular reports or maintain dashboards of their metrics.
```

The package has been designed to support several common use cases:

1. **Educational Analytics**: Tracking engagement across course materials hosted on various platforms
2. **Research Impact**: Monitoring repository activity and documentation usage
3. **Event Management**: Analyzing registration and attendance patterns through Calendly integration

> 4. **Website Performance**: Collecting and analyzing Google Analytics metrics

Reviewer: ```suggestion
4. **Website Engagement**: Collecting and analyzing Google Analytics metrics
```

> 5. **Community Engagement**: Measuring interaction through YouTube statistics and Google Forms responses

Reviewer: ```suggestion
5. **Community Interaction**: Measuring interaction through YouTube statistics and Google Forms responses
```

# Implementation and Features

The package implements several key features that facilitate metrics collection:

> 1. **Metric Standardization**: Large and unweildy json files are generally returned from most REST APIs. However this can be intimidating for many users to wrangle into a format that is easier to report or make data visualizations. All metrics returned from  `metricminer` are by default returned in tidy data format [@wickham_tidy_2014] , making it immediately usable with popular R packages like `dplyr` and `ggplot2`. For cases where the default transformation doesn't meet specific needs, the package provides an argument that will return instead the raw API responses. Ongoing community efforts to understand what metrics should be standard in reports can be facilitated as a part of `metricminer`.

Reviewer: ```suggestion
1. **Metric Standardization**: Large and unweildy json files are generally returned from most REST APIs. However this can be intimidating for many users to wrangle into a format that is easier to report or suitable for making data visualizations. All metrics returned from  `metricminer` are by default returned in tidy data format [@wickham_tidy_2014] , making it immediately usable with popular R packages like `dplyr` and `ggplot2`. For cases where the default transformation doesn't meet specific needs, the package provides an argument that will return instead the raw API responses. Ongoing community efforts to understand what metrics should be standard in reports can be facilitated as a part of `metricminer`.
```

# Example usage

The authorize function takes care of all API authorization for any apps the user is attempting to collect metrics from.

```
library(metricminer)
authorize("github")
```

By running this code it will trigger some menu options:
```
Would you like to store/cache your credentials?

1: Yes cache/store credentials
2: No do not store credentials, I will re-run this authorize() in my next R session
```
Credentials have the option to be cached and can be deleted anytime by running the `delete_creds()` function.

```
You chose to cache your credentials, if you change your mind, run metricminer::delete_creds().
Be careful not to push the cache files to GitHub or share it anywhere.
```

> For authorization, the proper pages will be open and instructions given for the user to provide their credentials.

Reviewer: ```suggestion
For authorization, the proper pages will be opened and instructions given for the user to provide their credentials.
```

```
On the opened page, scroll down and click 'Generate Token'.
```

The authorization process can be done for GitHub, Google, or Calendly.

> After authorization then you can run individual metric mining functions like:

Reviewer: ```suggestion
After authorization, then you can run individual metric mining functions like:
```

```
metrics <- get_github_repo_timecourse(
  repo = "fhdsl/metricminer"
  )
```

This will return a data frame stored in `metrics` that will look something like this.

|repo|timestamp|count_clones|uniques_clones|count_views|uniques_views|
|---|---|---|---|---|---|
|fhdsl/metricminer-dashboard|2024-04-29|1|2|1|1|
|fhdsl/metricminer-dashboard|2024-04-30|1|1|5|1|

> |fhdsl/metricminer-dashboard|2024-05-01|1|1|1|NA|

Reviewer: Does the NA in the third row, final column mean that no data was available for that day? Why would there be a view (in the third column) if there isn't a unique view? Is this just an example that I'm overthinking?

Author: Yeah it can be possible. Because unique_views can only be counted when a person is logged in. 

That being said, so readers aren't distracted by this, I'm all for just artificially changing this to 1. 

> Why uniques can be 0 or NA when count is 1
> 1. Privacy threshold / anonymization
GitHub applies a minimum threshold for unique visitor tracking. If a visitor's identity cannot be reliably determined (e.g., not logged in, bot-like behavior, privacy settings), GitHub may record the view (count) but not attribute it to a unique visitor (uniques).
> 2. Unauthenticated/anonymous traffic
Logged-out users are counted in count but may not contribute to uniques if GitHub can't distinguish them from other anonymous sessions.
> 3. Bot/crawler traffic
Automated requests (scrapers, crawlers, CI pipelines fetching README badges) increment count but GitHub may exclude them from uniques.
> 4. API/data lag
There can be a short window where count has propagated but uniques hasn't been computed yet, leading to a temporary mismatch.

Reviewer 2: I do think this is useful to share with people, but maybe in our documentation instead of the paper? Seems like it might be a detail that could could be distracting like you said.

Reviewer 2: changed to a 1

---
title: 'Metricminer:'
tags:
  - R
  - API
  - Google
  - Calendly
  - Slido
  - GitHub
authors:
  - name: Candace Savonen
    orcid: 0000-0001-6331-7070
    corresponding: true
    affiliation: 1
  - name: Kate Isaac
    affiliation: 1

>   - name: Howard Baek

Reviewer: New spelling for Howard: https://www.linkedin.com/in/howard-baik/

Reviewer 3: Hi! Hope all is well. Thanks for catching this. I changed my legal name to “Howard Baik”.

>   - name: Howard Baek

Reviewer: ```suggestion
  - name: Howard Baik
```

Reviewer 3: Hi! Hope all is well. Thanks for catching this. I changed my legal name to “Howard Baik”.

    affiliation: 2
  - name: Carrie Wright
    affiliation: 1
  - name: Jeff Leek
    affiliation: 1
affiliations:
 - name: Fred Hutchinson Cancer Center, Seattle, WA, United States
   index: 1

>  - name: Yale School of Public Health

Reviewer: New Affiliation for Howard: https://www.linkedin.com/in/howard-baik/

>  - name: Yale School of Public Health

Reviewer: ```suggestion
 - name: Integral Consulting, Boulder, CO, United States
```

   index: 2
date: 12 March 2025
bibliography: paper.bib
---

# Summary

> The collection and analysis of standardized metrics is crucial for understanding the true impact and usage patterns of research software, extending far beyond traditional citation counts [@bioinformatics]. Citations alone fail to capture many critical aspects of software utilization, including version information, indirect usage through platforms, and evolving impact over time. "Metric mining" aims to systematically collect and analyze comprehensive usage data, making it accessible to developers, funders, and evaluation committees. There is a vital need for automated, standardized metric collection so that software impact can be demonstrated to funders and potential users. `metricminer` is an open source R package to help address the difficulty and lack of standardization for metric collection. 

Reviewer: @cansavvy Do we still need to add additional citations for this paragraph as you've referenced in this github comment from the old repo/PR? https://github.com/fhdsl/metricminer/pull/99#discussion_r1995712229

Author: Nah that's good.

# Statement of need

Previous research by our group showed that the main reason researchers do not collect software metrics is a lack of time and funding to do so [@bioinformatics]. This creates a compounding missed opportunity: metrics can help developers learn more about how their software is used and can help demonstrate impact to funders which would, in turn, give developers more insights and funding respectively to increase the software's impact. Yet, the lack of time to collect metrics means developers can't fully learn about their software or quickly communicate this to funders. Collecting meaningful metrics from Application Programming Interfaces (APIs) can be time consuming, so `metricminer` aims to be an open-source solution to make the metric collection process more efficient.

> `metricminer` is an R package designed to simplify the collection of metrics from common web services. R provides an ideal environment for data analysis and visualization while supporting reproducible research workflows. metricminer offers a consistent, user-friendly interface to a variety of web services, including GitHub, Google Analytics, Calendly, Google Forms, and YouTube. The package handles authentication, data retrieval, and transformation of raw API responses into tidy data frames, making results immediately useful for downstream analysis [@wickham_tidy_2014].

Reviewer 2: ```suggestion
`metricminer` is an R package designed to simplify the collection of metrics from common web services. R provides an ideal environment for data analysis and visualization while supporting reproducible research workflows. metricminer offers a consistent, user-friendly interface to a variety of web services, including GitHub, Google Analytics, Google Forms, and YouTube. The package handles authentication, data retrieval, and transformation of raw API responses into tidy data frames, making results immediately useful for downstream analysis [@wickham_tidy_2014].
```

Reviewer 2: @cansavvy are you ok with this and then we add a bit more to future plans including calendly?

Author: Sounds good.

`metricminer` was developed to serve researchers, educators, and other practitioners who need to collect and analyze metrics from multiple platforms. It has been specifically designed to support educational initiatives and research projects that require tracking engagement across various platforms. The combination of simplified authentication, consistent data formats, and integration with the R ecosystem makes `metricminer` particularly valuable for organizations that need to generate regular reports or maintain dashboards of their metrics.

The package has been designed to support several common use cases:

1. **Educational Analytics**: Tracking engagement across course materials hosted on various platforms
2. **Research Impact**: Monitoring repository activity and documentation usage

> 3. **Event Management**: Analyzing registration and attendance patterns through Calendly integration

Reviewer 2: do we have slido integration now? could mention that here and calendly as future direction for event management, also possible that people track attendance in a Google sheet

Reviewer 2: I thought we did when I was playing around with it, but now I am confused since it isn't mentioned, perhaps I am missing something

Reviewer: Yes, we have a working slido integration again/now (after the changes that made the old working one stop working for me) as described in the header here: https://github.com/ottrproject/metricminer/blob/main/R/slido.R. It assumes the slido exported results are in a shared google drive. You give the function the google drive name as opposed to an ID. And you'll have to authenticate for that drive so it's not as seamless as the other integrations

Reviewer: Added a PR here to update the vignette documentation to be in line with the current functionality: https://github.com/ottrproject/metricminer/pull/20

> 3. **Event Management**: Analyzing registration and attendance patterns through Calendly integration

Reviewer 2: ```suggestion
3. **Event Management**: Analyzing registration and attendance patterns through Google Sheet integration
```

```
Would you like to store/cache your credentials?

1: Yes cache/store credentials
2: No do not store credentials, I will re-run this authorize() in my next R session
```
Credentials have the option to be cached and can be deleted anytime by running the `delete_creds()` function.

```
You chose to cache your credentials, if you change your mind, run metricminer::delete_creds().
Be careful not to push the cache files to GitHub or share it anywhere.
```

For authorization, the proper pages will be opened and instructions given for the user to provide their credentials.

```
On the opened page, scroll down and click 'Generate Token'.
```

The authorization process can be done for GitHub, Google, or Calendly.
After authorization, then you can run individual metric mining functions like:

```
metrics <- get_github_repo_timecourse(
  repo = "fhdsl/metricminer"
  )
```

This will return a data frame stored in `metrics` that will look something like this.

|repo|timestamp|count_clones|uniques_clones|count_views|uniques_views|
|---|---|---|---|---|---|
|fhdsl/metricminer-dashboard|2024-04-29|1|2|1|1|
|fhdsl/metricminer-dashboard|2024-04-30|1|1|5|1|

> |fhdsl/metricminer-dashboard|2024-05-01|1|1|1|NA|

Reviewer 2: ```suggestion
|fhdsl/metricminer-dashboard|2024-05-01|1|1|1|1|
```
